src/OrbitQt/orbitmainwindow.cpp

```diff
@@ -1140,26 +1130,48 @@ bool OrbitMainWindow::eventFilter(QObject* watched, QEvent* event) {
   return QObject::eventFilter(watched, event);
 }
 
-void OrbitMainWindow::closeEvent(QCloseEvent* event) {
+bool OrbitMainWindow::ShowCloseWarningAndReturnResult() {
+  return QMessageBox::question(
+             this, QApplication::applicationName(),
+             "This discards any unsaved progress. Are you sure you want to continue?") ==
+         QMessageBox::Yes;
+}
+
+bool OrbitMainWindow::ShowCaptureAbortWarningAndReturnResult() {
+  return QMessageBox::question(this, "Capture in progress",
+                               "A capture is currently in progress. Do you want to abort the "
+                               "capture and exit Orbit?") == QMessageBox::Yes;
+}
+
+bool OrbitMainWindow::ExitRequestWithReturnCode(int return_code) {
   if (app_->IsCapturing()) {
-    event->ignore();
+    if (!ShowCaptureAbortWarningAndReturnResult()) return false;
 
-    if (QMessageBox::question(this, "Capture in progress",
-                              "A capture is currently in progress. Do you want to abort the "
-                              "capture and exit Orbit?") == QMessageBox::Yes) {
-      // We need for the capture to clean up - close as soon as this is done
-      app_->SetCaptureFailedCallback([&] { close(); });
-      app_->AbortCapture();
-    }
-  } else {
-    if (main_thread_executor_) {
-      main_thread_executor_->AbortWaitingJobs();
-    }
-    if (introspection_widget_ != nullptr) {
-      introspection_widget_->close();
-    }
-    QMainWindow::closeEvent(event);
+    // We need for the capture to clean up - exit as soon as this is done
+    app_->SetCaptureFailedCallback([return_code] { QApplication::exit(return_code); });
+    app_->AbortCapture();
+    return true;
+  }
+
+  if (!ShowCloseWarningAndReturnResult()) return false;
+
+  if (main_thread_executor_) {
+    main_thread_executor_->AbortWaitingJobs();
+  }
+  if (introspection_widget_ != nullptr) {
+    introspection_widget_->close();
+  }
```

Comment:
Both `main_thread_executor_` and `introspection_widget_ ` are smart pointers, but in one case there is implicit conversion, and in other - no. I think we should be consistent.

Reply:
Good catch. I compared to nullptr in both ifs.

Reply:
I guess that's my fault. I introduced the first one. Anton just moved it to a different location.